hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+            node.weights['depthwise'].update_precision(node.types['depthwise_t'].precision)
+
+            inferred_types.append('depthwise_t')
+
+        if 'pointwise_t' in types_to_infer:
+            pointwise_quantizer = node.get_attr('pointwise_quantizer', None)
+            if pointwise_quantizer is not None:
+                pointwise_width = pointwise_quantizer.bits
+                pointwise_integers = pointwise_quantizer.hls_type.integer
+                node.types['pointwise_t'].name = node.name + '_pointwise_t'
+                node.types['pointwise_t'].precision = pointwise_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'pointwise_t')
+                pointwise_width = node.types['pointwise_t'].precision.width
+                pointwise_integers = node.types['pointwise_t'].precision.integer
+            node.weights['pointwise'].update_precision(node.types['pointwise_t'].precision)
+
+            inferred_types.append('pointwise_t')
+        else:
+            pointwise_width = node.types['pointwise_t'].precision.width
+            pointwise_integers = node.types['pointwise_t'].precision.integer
+
+        if 'bias_t' in types_to_infer:
+            bias_quantizer = node.get_attr('bias_quantizer', None)
+            if bias_quantizer is not None:
+                bias_width = bias_quantizer.bits
+                bias_integers = bias_quantizer.hls_type.integer
+                node.types['bias_t'].name = node.name + '_bias_t'
+                node.types['bias_t'].precision = bias_quantizer.hls_type
+            else:
+                self._infer_default_type(node, 'bias_t')
+                bias_width = node.types['bias_t'].precision.width
+                bias_integers = node.types['bias_t'].precision.integer
+            node.weights['bias'].update_precision(node.types['bias_t'].precision)
+
+            inferred_types.append('bias_t')
+        else:
+            bias_width = node.types['bias_t'].precision.width
+            bias_integers = node.types['bias_t'].precision.integer
+
+        n_ops = node.get_attr('n_chan')
+        new_type = FixedPrecisionType(
+            width=int(max(np.ceil(input_width + pointwise_width + np.log2(n_ops)), bias_width) + 1),
+            integer=int(max(np.ceil(input_integers + pointwise_integers + np.log2(n_ops)), bias_integers) + 1),
+        )
+
+        if 'accum_t' in types_to_infer:
+            node.types['accum_t'].name = node.name + '_accum_t'
+            node.types['accum_t'].precision = new_type
+
+            inferred_types.append('accum_t')
+
+        if 'result_t' in types_to_infer:
+            node.types['result_t'].name = node.name + '_result_t'
+            node.types['result_t'].precision = new_type
+
+            inferred_types.append('result_t')
+
+        return inferred_types
+
+    def _infer_bn_precision(self, node, types_to_infer):
+        inferred_types = []
+
+        if 'scale_t' in types_to_infer:
```

Comment:
Wouldn't the input quantization in this case be the input mean and variance (+ other things), not the scale and bias?

Reply:
(actually I have to see how this is handled by the qkeras parser)